### ISSUE TITLE
New rule: no-hidden-type-assertions (`getMeAT<T>(): T`)

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#735]
From https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

> `getMeAT<T>(): T`: If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion. Prefer to use a real type assertion, e.g. `getMeAT() as number`. Example where a type parameter is acceptable: `function id<T>(value: T): T;`. Exception: `new Map<string, number>()` is OK.

Despite that guidance, lots of libraries do this, including Axios: https://github.com/danielnixon/eslint-config-typed-fp/issues/286

The rule we're proposing here should catch when such a naughty function is defined _and_ when it is called to catch usage of functions that are outside our power to fix (e.g. Axios).

**Comment by danielnixon** [comment]
The only safe way to call such a function is by providing `unknown` as the type param that affects the result, then parsing using something like io-ts, zod, etc.

**Comment by danielnixon** [comment]
The simple case of this is easy enough to analyse:

```ts
declare const simple: <T>() => T;
```

And so are cases where the generic type param appears in positions like this:

```ts
declare const simple: <T>() => Promise<T>;
```

But things start to get _interesting_ when you consider how expressive TypeScript is:

```ts
declare const getMeAT: <T>() => T extends 42 ? 43 : 45;
```

```ts
declare const getMeAT: <T extends { a: U }, U>() => T["a"];
```

We could accumulate all of those using `isTypeReferenceNode`, `isConditionalTypeNode`, `isIndexedAccessTypeNode` and so on.

But that wouldn't solve this:

```ts
declare const getMeAT: <T extends { a: U }, U>() => T extends Array<infer T> ? T : T;
```

In this example the `T` in `infer T` is a different type to the outer `T` (which corresponds to the type param).

I wonder if TypeScript provides something we can use to determine if a type param "contributed to" a return type and/or a param type.
